commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
+ *    </li>
+ * </ul>
+ *
+ * @since 8.0
+ */
+public final class Listeners {
+
+   private Listeners() {
+      // Cannot be instantiated, it's just a holder class
+   }
+
+   /**
+    * Read-write listeners enables user to register listeners for events
+    * happening in the read-write functional map.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Read-write listeners can distinguish between cache entry
+    *    creation versus cache entry modification or cache entry update
+    *    (a better name btw!). This is because it can read the previous
+    *    value and find out whether when the cache entry was written, whether
+    *    the previous cache entry was present or not. {@link WriteListeners}
+    *    cannot make such distinction because they are not allowed to read
+    *    the previous value.
+    *    </li>
+    *    <li>Read-write listeners can find out about entry removed events
+    *    and what the previous value for the entry was. Write-only listeners
+    *    do not have such capability since once again, they have no access to
+    *    previously stored data.
+    *    </li>
+    * </ul>
+    */
+   public interface ReadWriteListeners<K, V> {
+      /**
+       * Add a create event specific listener by passing in a
+       * {@link Consumer} to be called back each time a new cache entry is
+       * created, passing in a {@link ReadEntryView} of that entry.
+       *
+       * This method is shortcut for users who are only interested in
+       * create events. If interested in multiple event types, calling
+       * {@link #add(ReadWriteListener)} is recommended instead.
+       *
+       * @param f operation to be called each time a new cache entry is created
+       * @return an {@link AutoCloseable} instance that can be used to
+       *         unregister the listener
+       */
+      AutoCloseable onCreate(Consumer<ReadEntryView<K, V>> f);
```

Comment:
@danberindei Here's my POV: you need to find way to remove listeners. An easy way to do so is to return a handle that you can close/remove. With this in mind, you either: come up with a new API, or reuse what's out there. AutoCloseable looks very usable here IMO, offering both the possibility to close things yourself manually, e.g. call close() directly, or take advantage of the `try()` syntatic sugar. I've found the latter particularly useful in unit testing which I'll submit later today.